### PR TITLE
HIVE-25963: Create temporary table with not null constraint gets converted to external table

### PR DESCRIPTION
**Pull request by sourabh912**
### What changes were proposed in this pull request?

This PR fixes an issue in which a temporary table when created with not null column constraint, gets converted to external table.  It is because HMSClient api createTableWithConstraints() call createTable(createTableRequest) which is not overriden by SessionHiveMetaStoreClient (which handles temporary tables creation/fetch/deletion). As a result of which HMS client makes create_table() RPC call to HMS server which in turn changes the table to external table and also persists its metadata. This is not the desired behavior because temporary tables are stored in memory and the metadata for these should not persisted in HMS.   

### Why are the changes needed?

It fixes a bug as described in HIVE-25963

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Enhanced the existing unit tests by adding positive cases.

### CODE REVIEW

**Comment by sourabh912** [comment]
@nrg4878 : Please review and provide your feedback.

File: ql/src/test/results/clientpositive/llap/enforce_constraint_notnull.q.out [hunk 1]
```diff
@@ -6411,53 +6364,16 @@ STAGE PLANS:
                     expressions: UDFToInteger(key) (type: int)
                     outputColumnNames: _col0
                     Statistics: Num rows: 500 Data size: 2000 Basic stats: COMPLETE Column stats: COMPLETE
-                    Filter Operator
-                      predicate: enforce_constraint(_col0 is not null) (type: boolean)
```
Comment (hsnusonic) [inline ql/src/test/results/clientpositive/llap/enforce_constraint_notnull.q.out:6415]: Does this mean the not null constraint is not applied after the change?

**Comment by sourabh912** [comment]
The only test `stats_part_multi_insert_acid.q` failing in the latest commit seems unrelated to this patch because it is failing in other Hive patches too for example: http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3049/1/tests

File: ql/src/test/results/clientpositive/llap/enforce_constraint_notnull.q.out (hunk 1 above)
Comment (sourabh912) [reply ql/src/test/results/clientpositive/llap/enforce_constraint_notnull.q.out:6415]: Yes, it means that not null constraint does not have any meaning for temporary tables. Before this patch, a temp table with constraints was getting persisted in metastore and therefore had constraint information persisted. But  after this patch, constraints information will get lost for temporary tables unless we store constraints related info in memory. 

@nrg4878 : Is the behavior for temporary table in terms on constraints as expected after this patch?

**Comment by sourabh912** [comment]
@yongzhi @harishjp @saihemanth-cloudera @hsnusonic  : Would appreciate feedback from you as well !

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 2]
```diff
@@ -1247,31 +1247,11 @@ public void createTable(Table tbl) throws AlreadyExistsException,
 
   public void createTable(Table tbl, EnvironmentContext envContext) throws AlreadyExistsException,
       InvalidObjectException, MetaException, NoSuchObjectException, TException {
-    if (!tbl.isSetCatName()) {
-      tbl.setCatName(getDefaultCatalog(conf));
```
Comment (saihemanth-cloudera) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1251]: catalog name needs to be set if it is not set. Can you please add these lines back?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 3]
```diff
@@ -1247,31 +1247,11 @@ public void createTable(Table tbl) throws AlreadyExistsException,
...
-          hook.rollbackCreateTable(tbl);
-        } catch (Exception e){
-          LOG.error("Create rollback failed with", e);
-        }
-      }
+    CreateTableRequest request = new CreateTableRequest(tbl);
+    if (envContext != null) {
+      request.setEnvContext(envContext);
     }
+    createTable(request);
```
Comment (saihemanth-cloudera) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1254]: Can you please add L4448-4451 to the CreateTableRequest object here before calling the CreateTable() api? We would need to send the processor capabilities (if available) to HMS.

**Review by saihemanth-cloudera — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java (hunk 2 above)
Comment (sourabh912) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1251]: Good catch ! 
I will add them back.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java (hunk 3 above)
Comment (sourabh912) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1254]: `createTable(request)` method at line number 1284 is taking care of that already.

**Review by hsnusonic — APPROVED** [review]
Looks good to me!

**Comment by nrg4878** [comment]
Fix has been merged. Please close the PR and the jira as well.

**Pull request closed without merging**